Values exported using JsonSchemaConverter failing in sink connector

Hi,
I am trying to configure an azure blob sink connector to export the JSON records that are stored in a topic. The producer is producing the records using JsonSchema. Export works fine if there is no optional fields in the json schema. However if the json schema contains optional attributes and if they are not populated, the connector fails. 

Steps to reproduce the problem below:

**JSON Schema**
```
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "Test",
    "type": "object",
    "additionalProperties": false,
    "properties": {
        "id": {
            "type": "string",
            "description": "unique id"
        },
        "createById": {
            "type": "string",
            "description": "user id"
        },
        "complexNode": {
            "$ref": "#/definitions/complexNode"
        }
    },
    "required": [
        "id"
    ],
    "definitions": {
        "complexNode": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "id": {
                    "type": "string",
                    "description": "unique id"
                },
                "createById": {
                    "type": "string",
                    "description": "user id"
                }
            },
            "required": [
                "id"
            ]
        }
    }
}
```

**Produce Sample Message**

publish the following message to the topic

```
{"id":"xyz"}
```

**Kafka Connect Configuration**

set `value.converter=io.confluent.connect.json.JsonSchemaConverter` in the kafka-connect properties
set `format.class=io.confluent.connect.azure.blob.format.json.JsonFormat` in the azure blob sink properties.

**Error**
```
[2021-03-16 10:26:21,336] ERROR WorkerSinkTask{id=blob-sink-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted (org.apache.kafka.connect.runtime.WorkerTask:187)
org.apache.kafka.connect.errors.ConnectException: Exiting WorkerSinkTask due to unrecoverable exception.
	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:614)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:329)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:232)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:201)
	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:185)
	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:234)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.apache.kafka.connect.errors.DataException: Conversion error: null value for field that is required and has no default value
	at org.apache.kafka.connect.json.JsonConverter.convertToJson(JsonConverter.java:611)
	at org.apache.kafka.connect.json.JsonConverter.convertToJson(JsonConverter.java:704)
	at org.apache.kafka.connect.json.JsonConverter.convertToJsonWithoutEnvelope(JsonConverter.java:596)
	at org.apache.kafka.connect.json.JsonConverter.fromConnectData(JsonConverter.java:346)
	at io.confluent.connect.azure.storage.format.json.JsonRecordWriterProvider$1.write(JsonRecordWriterProvider.java:67)
	at io.confluent.connect.azure.storage.TopicPartitionWriter.writeRecord(TopicPartitionWriter.java:389)
	at io.confluent.connect.azure.storage.TopicPartitionWriter.checkRotationOrAppend(TopicPartitionWriter.java:209)
	at io.confluent.connect.azure.storage.TopicPartitionWriter.executeState(TopicPartitionWriter.java:172)
	at io.confluent.connect.azure.storage.TopicPartitionWriter.write(TopicPartitionWriter.java:141)
	at io.confluent.connect.azure.blob.AzureBlobStorageSinkTask.put(AzureBlobStorageSinkTask.java:195)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:586)
	... 10 more
```

I assume the root cause is - the `required` setting in the json schema is not being respected when converting to STRUCT schema? Wondering if there is any workaround to get around this issue?

@rayokota thanks for looking into the issue. You may be right..but I don't think the issue is fully resolved by #1754. I am using the latest version of confluent which includes #1754. If I run the connector using the previous version of the confluent, it fails in the `toConnectData()` with the following error

```
[2021-03-17 07:49:16,408] ERROR WorkerSinkTask{id=test-sink-sit-0} Task threw an uncaught and unrecoverable exception (org.apache.kafka.connect.runtime.WorkerTask:187)
org.apache.kafka.connect.errors.ConnectException: Tolerance exceeded in error handler
        at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:196)
        at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execute(RetryWithToleranceOperator.java:122)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.convertAndTransformRecord(WorkerSinkTask.java:495)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.convertMessages(WorkerSinkTask.java:472)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:322)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:226)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:198)
        at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:185)
        at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:235)
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: org.apache.kafka.connect.errors.DataException: Invalid null value for required STRING field
        at io.confluent.connect.json.JsonSchemaData.toConnectData(JsonSchemaData.java:511)
        at io.confluent.connect.json.JsonSchemaData.lambda$static$11(JsonSchemaData.java:192)
        at io.confluent.connect.json.JsonSchemaData.toConnectData(JsonSchemaData.java:561)
        at io.confluent.connect.json.JsonSchemaData.lambda$static$11(JsonSchemaData.java:192)
        at io.confluent.connect.json.JsonSchemaData.toConnectData(JsonSchemaData.java:561)
        at io.confluent.connect.json.JsonSchemaConverter.toConnectData(JsonSchemaConverter.java:108)
        at org.apache.kafka.connect.storage.Converter.toConnectData(Converter.java:87)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.lambda$convertAndTransformRecord$1(WorkerSinkTask.java:495)
        at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndRetry(RetryWithToleranceOperator.java:146)
        at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:180)
        ... 13 more
[2021-03-17 07:49:16,410] ERROR WorkerSinkTask{id=test-sink-sit-0} Task is being killed and will not recover until manually restarted (org.apache.kafka.connect.runtime.WorkerTask:188)
[2021-03-17 07:49:16,410] INFO Stopping SinkTask for test-sink-sit (io.confluent.connect.azure.blob.AzureBlobStorageSinkTask:262)
```

**After #1754 **
It is no longer failing in the `toConnectData()`. However if you look at the stack-trace, it is failing while converting from the connect data in the `fromConnectData`

```
[2021-03-16 10:26:21,336] ERROR WorkerSinkTask{id=blob-sink-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted (org.apache.kafka.connect.runtime.WorkerTask:187)
org.apache.kafka.connect.errors.ConnectException: Exiting WorkerSinkTask due to unrecoverable exception.
	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:614)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:329)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:232)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:201)
	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:185)
	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:234)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.apache.kafka.connect.errors.DataException: Conversion error: null value for field that is required and has no default value
	at org.apache.kafka.connect.json.JsonConverter.convertToJson(JsonConverter.java:611)
	at org.apache.kafka.connect.json.JsonConverter.convertToJson(JsonConverter.java:704)
	at org.apache.kafka.connect.json.JsonConverter.convertToJsonWithoutEnvelope(JsonConverter.java:596)
	at org.apache.kafka.connect.json.JsonConverter.fromConnectData(JsonConverter.java:346)
	at io.confluent.connect.azure.storage.format.json.JsonRecordWriterProvider$1.write(JsonRecordWriterProvider.java:67)
	at io.confluent.connect.azure.storage.TopicPartitionWriter.writeRecord(TopicPartitionWriter.java:389)
	at io.confluent.connect.azure.storage.TopicPartitionWriter.checkRotationOrAppend(TopicPartitionWriter.java:209)
	at io.confluent.connect.azure.storage.TopicPartitionWriter.executeState(TopicPartitionWriter.java:172)
	at io.confluent.connect.azure.storage.TopicPartitionWriter.write(TopicPartitionWriter.java:141)
	at io.confluent.connect.azure.blob.AzureBlobStorageSinkTask.put(AzureBlobStorageSinkTask.java:195)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:586)
	... 10 more
```

Hence I raised this issue.

--- a/json-schema-converter/src/main/java/io/confluent/connect/json/JsonSchemaData.java
+++ b/json-schema-converter/src/main/java/io/confluent/connect/json/JsonSchemaData.java
@@ -972,7 +972,9 @@ public class JsonSchemaData {
         for (Map.Entry<String, org.everit.json.schema.Schema> property : sortedMap.values()) {
           String subFieldName = property.getKey();
           org.everit.json.schema.Schema subSchema = property.getValue();
-          builder.field(subFieldName, toConnectSchema(subSchema));
+          boolean isFieldOptional = config.useOptionalForNonRequiredProperties()
+              && !objectSchema.getRequiredProperties().contains(subFieldName);
+          builder.field(subFieldName, toConnectSchema(subSchema, null, isFieldOptional));
         }
       }
     } else if (jsonSchema instanceof ReferenceSchema) {

--- a/json-schema-converter/src/main/java/io/confluent/connect/json/JsonSchemaDataConfig.java
+++ b/json-schema-converter/src/main/java/io/confluent/connect/json/JsonSchemaDataConfig.java
@@ -15,8 +15,6 @@
 
 package io.confluent.connect.json;
 
-import io.confluent.common.config.ConfigDef.Importance;
-import io.confluent.common.config.ConfigDef.Type;
 import org.apache.kafka.common.utils.Utils;
 import org.apache.kafka.connect.json.DecimalFormat;
 

--- a/json-schema-converter/src/main/java/io/confluent/connect/json/JsonSchemaDataConfig.java
+++ b/json-schema-converter/src/main/java/io/confluent/connect/json/JsonSchemaDataConfig.java
@@ -15,6 +15,8 @@
 
 package io.confluent.connect.json;
 
+import io.confluent.common.config.ConfigDef.Importance;
+import io.confluent.common.config.ConfigDef.Type;
 import org.apache.kafka.common.utils.Utils;
 import org.apache.kafka.connect.json.DecimalFormat;
 
@@ -32,6 +34,11 @@ import io.confluent.common.config.ConfigException;
 
 public class JsonSchemaDataConfig extends AbstractConfig {
 
+  public static final String USE_OPTIONAL_FOR_NON_REQUIRED_CONFIG = "use.optional.for.nonrequired";
+  public static final boolean USE_OPTIONAL_FOR_NON_REQUIRED_DEFAULT = false;
+  public static final String USE_OPTIONAL_FOR_NON_REQUIRED_DOC =
+      "Whether to set non-required properties to be optional.";
+
   public static final String SCHEMAS_CACHE_SIZE_CONFIG = "schemas.cache.size";
   public static final int SCHEMAS_CACHE_SIZE_DEFAULT = 1000;
   public static final String SCHEMAS_CACHE_SIZE_DOC = "Size of the converted schemas cache";
@@ -41,10 +48,15 @@ public class JsonSchemaDataConfig extends AbstractConfig {
   private static final String DECIMAL_FORMAT_DOC =
       "Controls which format this converter will serialize decimals in."
       + " This value is case insensitive and can be either 'BASE64' (default) or 'NUMERIC'";
-  private static final String DECIMAL_FORMAT_DISPLAY = "Decimal Format";
 
   public static ConfigDef baseConfigDef() {
     return new ConfigDef().define(
+        USE_OPTIONAL_FOR_NON_REQUIRED_CONFIG,
+        ConfigDef.Type.BOOLEAN,
+        USE_OPTIONAL_FOR_NON_REQUIRED_DEFAULT,
+        ConfigDef.Importance.MEDIUM,
+        USE_OPTIONAL_FOR_NON_REQUIRED_DOC
+    ).define(
         SCHEMAS_CACHE_SIZE_CONFIG,
         ConfigDef.Type.INT,
         SCHEMAS_CACHE_SIZE_DEFAULT,
@@ -63,6 +75,10 @@ public class JsonSchemaDataConfig extends AbstractConfig {
 
   public JsonSchemaDataConfig(Map<?, ?> props) {
     super(baseConfigDef(), props);
+  }
+
+  public boolean useOptionalForNonRequiredProperties() {
+    return getBoolean(USE_OPTIONAL_FOR_NON_REQUIRED_CONFIG);
   }
 
   public int schemaCacheSize() {

--- a/json-schema-converter/src/test/java/io/confluent/connect/json/JsonSchemaDataTest.java
+++ b/json-schema-converter/src/test/java/io/confluent/connect/json/JsonSchemaDataTest.java
@@ -33,6 +33,7 @@ import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import java.util.ArrayList;
 import java.util.List;
+import org.apache.avro.data.Json;
 import org.apache.kafka.connect.data.ConnectSchema;
 import org.apache.kafka.connect.data.Decimal;
 import org.apache.kafka.connect.data.Schema;
@@ -743,6 +744,36 @@ public class JsonSchemaDataTest {
   }
 
   @Test
+  public void testToConnectRecordWithOptionalAndRequired() {
+    NumberSchema numberSchema = NumberSchema.builder()
+        .requiresInteger(true)
+        .unprocessedProperties(ImmutableMap.of("connect.index", 0, "connect.type", "int8"))
+        .build();
+    StringSchema stringSchema = StringSchema.builder()
+        .unprocessedProperties(ImmutableMap.of("connect.index", 1))
+        .build();
+    ObjectSchema schema = ObjectSchema.builder()
+        .addPropertySchema("int8", numberSchema)
+        .addPropertySchema("string", stringSchema)
+        .addRequiredProperty("int8")
+        .title("Record")
+        .build();
+    ObjectNode obj = JsonNodeFactory.instance.objectNode();
+    obj.set("int8", ShortNode.valueOf((short) 12));
+    Schema expectedSchema = SchemaBuilder.struct()
+        .name("Record")
+        .field("int8", Schema.INT8_SCHEMA)
+        .field("string", Schema.OPTIONAL_STRING_SCHEMA)
+        .build();
+    Struct struct = new Struct(expectedSchema).put("int8", (byte) 12);
+    JsonSchemaDataConfig jsonSchemaDataConfig = new JsonSchemaDataConfig.Builder()
+        .with(JsonSchemaDataConfig.USE_OPTIONAL_FOR_NON_REQUIRED_CONFIG, true)
+        .build();
+    JsonSchemaData jsonSchemaData = new JsonSchemaData(jsonSchemaDataConfig);
+    checkNonObjectConversion(jsonSchemaData, expectedSchema, struct, schema, obj);
+  }
+
+  @Test
   public void testToConnectRecordWithMissingNonoptional() {
     NumberSchema numberSchema = NumberSchema.builder()
         .requiresInteger(true)
@@ -1244,6 +1275,13 @@ public class JsonSchemaDataTest {
   }
 
   private void checkNonObjectConversion(
+      Schema expectedSchema, Object expected, org.everit.json.schema.Schema schema, JsonNode value
+  ) {
+    checkNonObjectConversion(jsonSchemaData, expectedSchema, expected, schema, value);
+  }
+
+  private void checkNonObjectConversion(
+      JsonSchemaData jsonSchemaData,
       Schema expectedSchema, Object expected, org.everit.json.schema.Schema schema, JsonNode value
   ) {
     Schema connectSchema = jsonSchemaData.toConnectSchema(schema);

--- a/json-schema-converter/src/test/java/io/confluent/connect/json/JsonSchemaDataTest.java
+++ b/json-schema-converter/src/test/java/io/confluent/connect/json/JsonSchemaDataTest.java
@@ -33,7 +33,6 @@ import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import java.util.ArrayList;
 import java.util.List;
-import org.apache.avro.data.Json;
 import org.apache.kafka.connect.data.ConnectSchema;
 import org.apache.kafka.connect.data.Decimal;
 import org.apache.kafka.connect.data.Schema;

